enhancements/kube-apiserver/bound-sa-tokens.md

```diff
@@ -0,0 +1,146 @@
+
+## Summary
+
+Enable the optional use of bound service account tokens via volume projection and the
+TokenRequest API.
+
+## Motivation
+
+3rd party IAM components such as the [AWS pod identity
+webhook](https://github.com/aws/amazon-eks-pod-identity-webhook) require bound tokens to
+be able to identify pods without the use of workarounds such as proxies. It is also
+generally desirable to support bound tokens to limit the scope of permissions (and
+therefore risk of compromise) for a given service account token.
+
+### Goals
+
+- A pod can request a bound service account token via a projected volume.
+- A bound token can be acquired via the TokenRequest API.
+- The public key used to validate bound tokens can be retrieved for use with 3rd party IAM
+  components.
+
+### Non-Goals
+
+- Integrate or ship the aws pod identity webhook
+- Convert existing token-consuming components to use bound tokens
+
+## Proposal
+
+- Support configuring the options (via `KubeAPIServerConfig`) that will enable bound
+  tokens.
+  - The `TokenRequest` and `TokenRequestProjection` feature gates are enabled by default
+    in kube > 1.12. They are only configured by the apiserver if the following options are
+    provided:
+    - service-account-signing-key-file (set to private key managed by operator)
+    - service-account-issuer (defaults to https://kubernetes.default.svc/)
+    - api-audiences (defaults to issuer)
+
+- The apiserver operator should manage a keypair used to sign and verify bound tokens
+  - The keypair needs to be distinct from the keypair used for legacy service account
+    tokens. Reusing the legacy token keypair would prevent token validation since tokens
+    are first validated by key and then by content. If the key were to validate, the
+    subsequent content check would fail due to differences in the content of legacy and
+    bound tokens.
+  - The keypair should be written to a secret in the `openshift-kube-apiserver` namespace
+  - The public key should be added to (rather than replaced in) a configmap in the
+    `openshift-config-managed` namespace.
+    - This configmap can be used to source the public keys needed by 3rd party components
+      to verify bound tokens issued by the apiserver.
+    - The public keys in the configmap should be included in
+      `KubeAPIServerConfig.ServiceAccountPublicKeyFiles` for provision to the apiserver.
+      - In the event that the keypair secret is deleted, a new keypair will be generated
+        but tokens signed by the previous private key will continue to validate since the
+        corresponding public key will still be used to validate bound tokens.
+    - (Optional) Public keys known to be older than the maximum expiry of previously
+      issued tokens could be removed from the configmap.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Bound tokens are not currently required until bootstrap is complete, so it should be
```

Comment:
confirm that @abhinavdahiya doesn't want these for machines past the bootstrap node.

Reply:
> confirm that @abhinavdahiya doesn't want these for machines past the bootstrap node.

Where "want" is an actual product need. It's generally difficult to wire these twice pre- and post- operator

Reply:
services running on the bootstrap-host can use the ec2metadata to fetch temporary credentials so I don't think there is a desire today to support the tokens using the bootstrap-control-plane for controllers running on the bootstrap-host.

I'm hoping that's what @deads2k was referring.

As for one time (or expiring) tokens for bootstrapping nodes, that would be a interesting outcome, but that would not work if we want kubelets to rejoin after being shutdown for long enough that they need to use the bootstrap-kubeconfig to get kubeconfig again...

So I do not think it's a use-case yet.